### PR TITLE
Ignore clicks not directed at Plottable graphs. Fixes #2142

### LINKED ISSUE #2142: Click Interactions work behind overlays
Interactions can be detected through overlays:
http://jsfiddle.net/ohk89gth/

**Actual**: Clicking anywhere in the `<svg>` triggers a message.
**Expected**: Clicking on the part of the `<svg>` covered by the overlay doesn't trigger anything.

### PR DESCRIPTION
**Pull request by raiju**
Fixes #2142

### CODE REVIEW

File: src/dispatchers/mouseDispatcher.ts [hunk 1]
```diff
@@ -178,11 +178,13 @@ export module Dispatchers {
      * Computes the mouse position from the given event, and if successful
      * calls all the callbacks in the provided callbackSet.
      */
-    private _measureAndDispatch(event: MouseEvent, callbackSet: Utils.CallbackSet<MouseCallback>) {
-      var newMousePosition = this._translator.computePosition(event.clientX, event.clientY);
-      if (newMousePosition != null) {
-        this._lastMousePosition = newMousePosition;
-        callbackSet.callCallbacks(this.lastMousePosition(), event);
+    private _measureAndDispatch(event: MouseEvent, callbackSet: Utils.CallbackSet<MouseCallback>, captureAllEvents: boolean = false) {
```
Comment (dbt) [inline src/dispatchers/mouseDispatcher.ts]: Please don't use boolean parameters in methods.

http://blogs.msdn.com/b/oldnewthing/archive/2006/08/28/728349.aspx

**Comment by raiju** [comment]
Would a scope: string with "PAGE"/"ELEMENT" (or "CAPTURE_PAGE" and "CAPTURE_ELEMENT") be preferable? Or an enum? (In that case, where do you want the enum to live?)

**Comment by dbt** [comment]
In general we're using lowercase strings for these, and if it's a public method you should expose them as constants.  You don't need to do that here.  I'd lean towards "page"/"element" over capture_...

File: src/utils/clientToSVGTranslator.ts [hunk 2]
```diff
@@ -70,6 +70,18 @@ export module Utils {
 
       return scaledPosition;
     }
+    public isWithinTarget(e: MouseEvent) {
+      return this._isChildOf(<Element>e.target, this._svg);
```
Comment (jtlan) [inline src/utils/clientToSVGTranslator.ts]: You could probably use an existing `Utils` method here:
```Typescript
return Utils.DOM.boundingSVG(<SVGElement>e.target) === this._svg);
```

File: src/dispatchers/mouseDispatcher.ts [hunk 3]
```diff
@@ -178,11 +178,13 @@ export module Dispatchers {
      * Computes the mouse position from the given event, and if successful
      * calls all the callbacks in the provided callbackSet.
      */
-    private _measureAndDispatch(event: MouseEvent, callbackSet: Utils.CallbackSet<MouseCallback>) {
-      var newMousePosition = this._translator.computePosition(event.clientX, event.clientY);
-      if (newMousePosition != null) {
-        this._lastMousePosition = newMousePosition;
-        callbackSet.callCallbacks(this.lastMousePosition(), event);
+    private _measureAndDispatch(event: MouseEvent, callbackSet: Utils.CallbackSet<MouseCallback>, scope: string = "element") {
```
Comment (dbt) [inline src/dispatchers/mouseDispatcher.ts]: One tiny nit.  We should validate the input is "element" or "page".

**Comment by dbt** [comment]
lgtm

File: src/utils/clientToSVGTranslator.ts [hunk 4]
```diff
@@ -70,6 +70,9 @@ export module Utils {
 
       return scaledPosition;
     }
+    public isWithinTarget(e: MouseEvent) {
```
Comment (jtlan) [inline src/utils/clientToSVGTranslator.ts]: Probably should be renamed to "`insideSVG()`", and take a generic `Event`. (Renaming it also makes the method self-documenting).

File: src/dispatchers/mouseDispatcher.ts (hunk 3 above)
Comment (jtlan) [inline src/dispatchers/mouseDispatcher.ts]: Looks like this has been addressed.

**Comment by crmorford** [comment]
Tested:
- [x] click
- [x] panzoom
- [ ] double click
- [ ] drag

**Comment by jtlan** [comment]
@crmorford 
Two NOTEs probably have the same root cause. I'd expect no `Interaction`s to trigger on top of the modal, so there's probably a few corrections to be made here.

**Comment by crmorford** [comment]
Ok, I'm going to QE -1 then and pass this back to @raiju.

**Comment by bluong** [comment]
@jtlan Note that both the NOTEs occur on mouseover/out rather than on clickDown/clickUp

**Comment by jtlan** [comment]
Taking care of `Pointer` and `Key` probably requires some mechanism where they don't respond to `Dispatcher` updates if the dispatched event isn't targeted at the `<svg>`, which means that the `Dispatcher` probably needs to pass that information to its listeners somehow.

**Comment by crmorford** [comment]
Can I get a summary of what happened?
Tests were added, yes? And pointer/key weren't fixed. Any other functionality changes?

**Comment by raiju** [comment]
Oh, sorry. Yes, I added a test which tests whether events on an overlay above the SVG are captured by the underlying SVG. No other functionality changes were made though.

**Comment by crmorford** [comment]
Wow, thanks for the quick response :)

**Pull request merged**